talk/tools/profiling.tex

```diff
@@ -4,39 +4,228 @@
   \frametitle{Profiling}
   \begin{block}{Conceptually}
     \begin{itemize}
-      \item take a measurement of a performance aspect of a program
+      \item Take a measurement of a performance aspect of a program
       \begin{itemize}
-        \item where in my code is most of the time spent?
-        \item is my program compute or memory bound?
-        \item does my program make good use of the cache?
-        \item is my program using all cores most of the time?
-        \item how often are threads blocked and why?
-        \item which API calls are made and in which order?
+        \item Where in my code is most of the time spent?
+        \item Is my program compute or memory bound?
+        \item Does my program make good use of the cache?
+        \item Is my program using all cores most of the time?
+        \item How often are threads blocked and why?
+        \item Which API calls are made and in which order?
         \item ...
       \end{itemize}
-      \item the goal is to find performance bottlenecks
-      \item is usually done on a compiled program, not on source code
+      \item The goal is to find performance bottlenecks
+      \item Usually done on a compiled program, not on source code
     \end{itemize}
   \end{block}
 \end{frame}
 
 \begin{frame}[fragile]
-  \frametitle{perf, VTune and uProf}
-  \begin{block}{perf}
+  \frametitle{\mintinline{bash}{perf} -- Performance analysis tools for Linux}
+  \setlength{\leftmargini}{0pt}
     \begin{itemize}
-      \item perf is a powerful command line profiling tool for linux
-      \item compile with \mintinline{bash}{-g -fno-omit-frame-pointer}
-      \item \mintinline{bash}{perf stat -d <prg>} gathers performance statistics while running \mintinline{bash}{<prg>}
-      \item \mintinline{bash}{perf record -g <prg>} starts profiling \mintinline{bash}{<prg>}
-      \item \mintinline{bash}{perf report} displays a report from the last profile
-      \item More information in \href{https://perf.wiki.kernel.org/index.php/Main_Page}{this wiki}, \href{https://www.brendangregg.com/linuxperf.html}{this website} or \href{https://indico.cern.ch/event/980497/contributions/4130271/attachments/2161581/3647235/linux-systems-performance.pdf}{this talk}.
+      \item Powerful command line profiling tool for Linux
+      \item Not portable, the source code is part of the Linux kernel itself
+      \item Much lower overhead compared with \mintinline{bash}{valgrind}
+      \item In order to profile your code, make sure to compile with
+            \texttt{CXXFLAGS="-O2 -g -fno-omit-frame-pointer"}
```

Comment:
why `-O2` ? Should work with any level, no ?

Reply:
Yes, should work with all levels. I asked to point out that you should compile with optimizations enabled, see: https://github.com/hsf-training/cpluspluscourse/pull/381#discussion_r1010271904